### ISSUE TITLE
Question about SASRec

### ISSUE THREAD

**Issue by sooonwoo** [RUCAIBox/RecBole#1290]
Hi thanks for your impressive work! your implementation helps me a lot. 

I am working on SASRec and do experiment on Amazon dataset(sports, beauty).
It seems default loss type of SASRec is 'CE' and I can get promising results by using default setting.
* [Amazon-sports] NDCG@10: 0.0227 / HR@10: 0.048 
* [Amazon-beauty] NDCG@10: 0.0416 / HR@10: 0.083
 
However, the performance is much higher than reported on several papers(e.g. https://github.com/RuihongQiu/DuoRec). 
*[Amazon-sports] NDCG@10: 0.0169 / HR@10: 0.0336
* [Amazon-beauty] NDCG@10: 0.0281 / HR@10: 0.0627

Using 'CE' instead of 'BPR' is not convention in sequential recommendation benchmark? Did I miss something?

**Comment by Wicknight** [comment]
@sooonwoo Hello, thanks for your attention to RecBole!

First, I didn't find the parameter settings of the SASRec model in these papers, so I'm not sure whether they used the default settings in the experiment. And even in the multiple papers mentioned in the [links](https://github.com/RuihongQiu/DuoRec) you provide, these results are not uniform, so I speculate that these papers may use different parameter settings in the experiment, such as learning rate or random seed. Therefore I don't think you need to worry too much about this problem. I see that the code you provide is based on RecBole, so I think you can experiment with models mentioned in these papers under his open source code to check these results.

Second, for most sequential recommendation models, both BPR loss and cross entropy loss can be used. And for SASRec model, the loss function used in the original paper is the cross entropy loss, so the default setting has no problem.

**Comment by sooonwoo** [comment]
Appreciate for your thoughtful reply :)

I think if benchmark result table on several datasets and implemented methods is provided, it could help users to increase their confidence in their own experimental results!

**Comment by Wicknight** [comment]
Thank you for your advice and we'll consider it!

**Comment by zfan20** [comment]
> @sooonwoo Hello, thanks for your attention to RecBole!
> 
> First, I didn't find the parameter settings of the SASRec model in these papers, so I'm not sure whether they used the default settings in the experiment. And even in the multiple papers mentioned in the [links](https://github.com/RuihongQiu/DuoRec) you provide, these results are not uniform, so I speculate that these papers may use different parameter settings in the experiment, such as learning rate or random seed. Therefore I don't think you need to worry too much about this problem. I see that the code you provide is based on RecBole, so I think you can experiment with models mentioned in these papers under his open source code to check these results. Meanwhile, I find that some people have raised [similar questions](https://github.com/RuihongQiu/DuoRec/issues/3) about these papers, I think you can pay attention to it.
> 
> Second, for most sequential recommendation models, both BPR loss and cross entropy loss can be used. And for SASRec model, 1:1 negative sampling is used in training process, and then binary cross entropy is used for optimization according to original paper. So I think our default setting has no problem.

I would like to bring up this issue again. Actually, the default setting in Recbole is not binary cross entropy at all. It utilizes all items as negatives. Using all items as negatives will significantly improve the performance, even if I am not sure why it is that yet. Moreover, the rule of thumb for most previous works is to use only one negative, including BPR and BCE. I believe Recbole should double check this potential issue and make sure all algorithms are using the same full-item set as negatives or consistently using one negative item. Thanks!

**Comment by Sherry-XLL** [comment]
Hello @zfan20, thanks for your attention and valuable suggestions to RecBole! As you mentioned, CE rather than BCE loss function is implemented in RecBole, because we provide a unified function interface for all sequence models.

Specifically, we support two kinds of loss functions for sequential models, i.e., Cross Entropy (CE) loss and Bayesian Personalized Ranking (BPR) loss.

The parameter `loss_type (str)` decides the type of loss function.
- If it is set to `'CE'`, the training task is regarded as a multi-classification task and the target item is the ground truth. In this way, negative sampling is not needed.
- If it is set to `'BPR'`, the training task will be optimized in the pair-wise way, which maximizes the difference between the positive item and the negative one. In this way, negative sampling is necessary, such as setting `--train_neg_sample_args="{'distribution':'uniform', 'sample_num': 1}"`.

By default, we set the cross entropy loss function and all items are utilized as negatives, and negative sampling is not allowed when `loss_type: CE`. You can refer to our [documentation](https://recbole.io/docs/user_guide/model/sequential/sasrec.html) for more details.

To ensure the comparability of models, we consistently provide CE and BPR loss function interfaces for all sequential models. The frequently mentioned binary cross entropy loss function (BCE) will also be included in the future development plan. Please comment or open a new issue if you're still having questions.

Related issues: [#788](https://github.com/RUCAIBox/RecBole/issues/788), [#955](https://github.com/RUCAIBox/RecBole/issues/955), [#1014](https://github.com/RUCAIBox/RecBole/issues/1014), [#1055](https://github.com/RUCAIBox/RecBole/discussions/1055), [#1133](https://github.com/RUCAIBox/RecBole/issues/1133), [#1252](https://github.com/RUCAIBox/RecBole/discussions/1252), [#1290](https://github.com/RUCAIBox/RecBole/issues/1290), [#1333](https://github.com/RUCAIBox/RecBole/issues/1333), [#1626](https://github.com/RUCAIBox/RecBole/issues/1626)

Best wishes!
